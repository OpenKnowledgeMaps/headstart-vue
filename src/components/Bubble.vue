<template>
    <g class="bubble_frame" :transform="translateString" :id="'bubble' + id">
        <circle
                :class="{ area: !zoomedOut, zoom_selected: zoomedOut }"
                :r="r"
                cx="0"
                cy="0"
                style="fill-opacity: 0.8; display: block;"
                @mouseover="$emit('bubbleMouseEnter', id)"
                @mouseout="$emit('bubbleMouseLeave', id)"
                @click="$emit('bubbleClick', id)"
                @dblclick.prevent="$emit('bubbleDoubleClick', id)"
        >
            <title>{{title}}</title>
        </circle>
        <foreignObject
                id="area_title_object"
                class="headstart"
                :x="foreignObjectX"
                :y="foreignObjectY"
                :width="inscribedSquareLength"
                :height="inscribedSquareLength"
                style="visibility: visible; display: block;"
                @mouseover="$emit('bubbleMouseEnter', id)"
                @mouseout="$emit('bubbleMouseLeave', id)"
                @click="$emit('bubbleClick', id)"
                @dblclick.prevent="$emit('bubbleDoubleClick', id)"
        >
            <body style="word-wrap: break-word;">
            <div id="area_title" :style="{width: inscribedSquareLength + 'px', height: inscribedSquareLength + 'px'}">
                <h2 class="highlightable" style="font-size: 12px;">
                    {{title}}
                </h2>
            </div>
            </body>
        </foreignObject>
    </g>
</template>

<script>
  export default {
    name: "Bubble",
    props: {
      id: Number,
      x: Number,
      y: Number,
      r: Number,
      title: String,
      zoomedOut: Boolean
    },
    computed: {
      translateString: function() {
        return "translate(" + this.x + "," + this.y + ")";
      },
      inscribedSquareLength: function () {
        return Math.sqrt(2*this.r*this.r);
      },
      foreignObjectX: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      },
      foreignObjectY: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      }
    }
  }
</script>