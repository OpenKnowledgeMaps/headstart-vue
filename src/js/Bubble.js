export default {
  name: 'Bubble',
  props: {
    id: Number,
    x: Number,
    y: Number,
    r: Number,
    order: Number,
    area: String,
    hovered: Boolean,
    zoomed: Boolean,
    selected: Boolean,
  },
  computed: {
    translateString() {
      return `translate(${this.x},${this.y})`;
    },
    inscribedSquareLength() {
      return Math.sqrt(2 * this.r * this.r);
    },
    foreignObjectX() {
      return -0.5 * Math.sqrt(2 * this.r * this.r);
    },
    foreignObjectY() {
      return -0.5 * Math.sqrt(2 * this.r * this.r);
    },
  },
};
