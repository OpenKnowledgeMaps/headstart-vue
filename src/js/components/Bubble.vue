<template src="../../templates/bubble.html"></template>

<script>
  export default {
    name: "Bubble",
    props: {
      id: Number,
      x: Number,
      y: Number,
      r: Number,
      order: Number,
      area: String,
      hovered: Boolean
    },
    computed: {
      translateString: function() {
        return "translate(" + this.x + "," + this.y + ")";
      },
      inscribedSquareLength: function () {
        return Math.sqrt(2*this.r*this.r);
      },
      foreignObjectX: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      },
      foreignObjectY: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      }
    }
  }
</script>
