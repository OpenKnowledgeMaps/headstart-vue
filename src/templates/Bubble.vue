<script src="../js/Bubble.js">
</script>

<template>
<g class="bubble_frame" :transform="translateString" :id="'bubble' + id">
    <circle
            :class="{ 
                area: (!hovered && !(zoomed && selected)), 
                zoom_selected: zoomed && selected}"
            style="fill-opacity: 0.8; display: block;"
            :r="r"
            cx="0"
            cy="0"
            @mouseover="$emit('mouseEnter', id)"
            @mouseout="$emit('mouseLeave', id)"
            @click="$emit('click', id)"
            @dblclick.prevent="$emit('doubleClick', id)"
    >
        <title>{{area}}</title>
    </circle>
    <foreignObject
            id="area_title_object"
            class="headstart"
            :x="foreignObjectX"
            :y="foreignObjectY"
            :width="inscribedSquareLength"
            :height="inscribedSquareLength"
            style="visibility: visible; display: block;"
            @mouseover="$emit('mouseEnter', id)"
            @mouseout="$emit('mouseLeave', id)"
            @click="$emit('click', id)"
            @dblclick.prevent="$emit('doubleClick', id)"
    >
        <body style="word-wrap: break-word;">
        <div id="areborder: 1px #000 solid;a_title" :style="{width: inscribedSquareLength + 'px', height: inscribedSquareLength + 'px'}">
            <h2 class="highlightable" style="font-size: 12px;" v-show="!hovered">
                {{area}}
            </h2>
        </div>
        </body>
    </foreignObject>
</g>
</template>