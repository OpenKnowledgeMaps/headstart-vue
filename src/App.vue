<template>
  <div id="app">
    <Chart :height="chartDimensions.height" :width="chartDimensions.width">
        <Bubble
                v-for="bubble in bubbles"
                :key="bubble.id"
                v-bind="bubble"
                v-on:mouseEnter="onBubbleMouseEnter"
                v-on:mouseLeave="onBubbleMouseLeave"
                v-on:click="onBubbleMouseClick"
                v-on:doubleClick="onBubbleDoubleClick"
        ></Bubble>
        <Paper v-for="paper in papers"
               :key="paper.id"
               v-bind="paper"
               v-on:mouseEnter="onPaperMouseEnter"
               v-on:mouseLeave="onPaperMouseLeave"
               v-on:click="onPaperMouseClick"
               v-on:doubleClick="onPaperDoubleClick"
        ></Paper>
    </Chart>
  </div>
</template>

<script>
import Chart from './templates/Chart.vue';
import Paper from './templates/Paper.vue';
import Bubble from './templates/Bubble.vue';

import backendData from './js/backendFakeAPIData.js';
import config from './js/config.js';

export default {
  name: 'app',
  components: {
    Chart,
    Paper,
    Bubble,
  },
  data() {
    return {
      chartDimensions: {
        width: config.svgWidth,
        height: config.svgHeight,
      },
      zoomState: {
        factor: 1,
        translate: {
          x: 0,
          y: 0,
        },
        zoomed: false,
      },
      zoomFactor: 1,
      zoomed: false,
      papers: backendData.papers,
      bubbles: backendData.bubbles,
    };
  },
  methods: {
    onBubbleMouseEnter(id) {
      this.bubbles.find(bubble => bubble.id === id).hovered = true;
    },
    onBubbleMouseLeave(id) {
      this.bubbles.find(bubble => bubble.id === id).hovered = false;
    },
    onBubbleMouseClick(id) {
      // eslint-disable-next-line
      console.log(`Clicked Bubble ${id} - not implemented yet`);
    },
    onBubbleDoubleClick(id) {
      // eslint-disable-next-line
      console.log(`Doubleclicked Bubble ${id} - not implemented yet`);
    },
    onPaperMouseEnter(id) {
      // eslint-disable-next-line
      console.log(`Hovered Paper ${id} - not implemented yet`);
    },
    onPaperMouseLeave(id) {
      // eslint-disable-next-line
      console.log(`Mouse left Paper ${id} - not implemented yet`);
    },
    onPaperMouseClick(id) {
      // eslint-disable-next-line
      console.log(`Clicked Paper ${id} - not implemented yet`);
    },
    onPaperDoubleClick(id) {
      // eslint-disable-next-line
      console.log(`Doubleclicked Paper ${id} - not implemented yet`);
    },
  },
};
</script>
