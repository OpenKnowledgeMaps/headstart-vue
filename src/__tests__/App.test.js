import App from '../App';
import Bubble from '../js/components/Bubble';
import { shallowMount, mount } from '@vue/test-utils';

describe('App', () => {
  it('matches a previously made snapshot', () => {
    const wrapper = shallowMount(App);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onBubbleMouseEnter if a bubbleMouseEnter signal is emitted', () => {
    const wrapper = mount(App);
    const bubbleWrapper = wrapper.find(Bubble);
    expect(bubbleWrapper.exists()).toBeTruthy();
    expect(bubbleWrapper.vm.hovered).toBe(false);
    bubbleWrapper.vm.$emit('mouseEnter', bubbleWrapper.vm.id);
    expect(bubbleWrapper.vm.hovered).toBe(true);
  });

  it('calls onBubbleMouseLeave if a bubbleMouseLeave signal is emitted', () => {
    const wrapper = mount(App);
    const bubbleWrapper = wrapper.find(Bubble);
    expect(bubbleWrapper.exists()).toBeTruthy();
    bubbleWrapper.vm.$emit('mouseenter', bubbleWrapper.vm.id);
    expect(bubbleWrapper.vm.hovered).toBe(true);
    bubbleWrapper.vm.$emit('mouseLeave', bubbleWrapper.vm.id);
    expect(bubbleWrapper.vm.hovered).toBe(false);
  });
});