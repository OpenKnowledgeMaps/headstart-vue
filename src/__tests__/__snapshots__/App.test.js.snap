// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches a previously made snapshot 1`] = `
<div id="app">
  <chart-stub></chart-stub>
</div>
`;
